### ISSUE TITLE
runCommand fails when using __structuredAttrs

### ISSUE THREAD

**Issue by lunik1** [NixOS/nixpkgs#213641]
### Describe the bug
The trivial `runCommand` builder (and friends) fails on a trivial derivation when `__structuredAttrs` is set.

### Steps To Reproduce
Steps to reproduce the behaviour:
1. Build the following derivation:
```nix
runCommand "foo" {__structuredAttrs = true;} ''. .attrs.sh; mkdir ''${outputs[out]}''
```
and get this error
```
error: builder for '/nix/store/vj7a0vhacspbvfjyngqf12iq8gj9vagz-foo.drv' failed with exit code 1;
       last 1 log lines:
       > /nix/store/9krlzvny65gdc8s7kpb6lkx8cd02c25b-default-builder.sh: line 1: /setup: No such file or directory
       For full logs, run 'nix log /nix/store/vj7a0vhacspbvfjyngqf12iq8gj9vagz-foo.drv'.
```

### Expected behavior
Derivation should build, akin to
```nix
runCommand "foo" {} ''mkdir $out''
```

### Additional context
There is a partial workaround seen [here](https://github.com/NixOS/nixpkgs/blob/52b7c0e7248799e0ab9906a4157ba783995b02d1/pkgs/build-support/references-by-popularity/default.nix#L4) where `builder` is set instead of `buildCommand`, however this leaves the utility functions defined in `setup.sh` unavailable to the builder.

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
output here
```

**Comment by abathur** [comment]
Idk why it would be the case, but the error itself sounds like the $stdenv var is empty when trying to run default-builder.sh:

https://github.com/NixOS/nixpkgs/blob/92247c104e4363e6dbb2ebb794cbfd7165b3031d/pkgs/stdenv/generic/default-builder.sh#L2

Not sure if this might be related to https://github.com/NixOS/nixpkgs/commit/238a6053c43f7ac2e3dcc3d3c7f29f1e0c0be589 or if that's a red herring.

**Comment by abathur** [comment]
I ended up here because someone reported a slightly different message on matrix (`/nix/store/6xg259477c90a229xwmb53pdfkn6ig3g-default-builder.sh: line 6: genericBuild: command not found` while rebuilding on NixOS) that sounds like it might also be caused by `$stdenv` being unset if `errexit` was enabled in the running script in your case but not in theirs.

That said, someone [suggested](https://matrix.to/#/!KqkRjyTEzAGRiZFBYT:nixos.org/$cdZ2o5bkqJQyk7wOhyMOlmaH9yj_gcHZt7vXhhdYEjA?via=nixos.org&via=matrix.org&via=tchncs.de) that they try verifying the store, and they eventually [reported](https://matrix.to/#/!KqkRjyTEzAGRiZFBYT:nixos.org/$UKCgdu-1ZIvw7cKbfXM143c5qAoQWHCcE--CKtgMoHo?via=nixos.org&via=matrix.org&via=tchncs.de) that repairing the store solved the problem for then.

I suspect this means their case isn't actually related to the issue here, but I decided to document it just in case it helps connect dots later.
